dolphin/db/sqlalchemy/api.py

```diff
@@ -309,25 +311,154 @@ def volume_get_all(context, marker=None, limit=None, sort_keys=None,
     return NotImplemented
 
 
+def _pool_get_query(context, session=None):
+    return model_query(context, models.Pool, session=session)
+
+
+def _pool_get(context, pool_id, session=None):
+    result = (_pool_get_query(context, session=session)
+              .filter_by(id=pool_id)
+              .first())
+
+    if not result:
+        LOG.error(exception.PoolNotFound(id=pool_id))
+
+    return result
+
+
 def pool_create(context, values):
     """Create a pool from the values dictionary."""
-    return NotImplemented
+    if not values.get('id'):
+        values['id'] = uuidutils.generate_uuid()
+
+    pool_ref = models.Pool()
+    pool_ref.update(values)
+
+    session = get_session()
+    with session.begin():
+        session.add(pool_ref)
+
+    return _pool_get(context,
+                     pool_ref['id'],
+                     session=session)
+
+
+def pools_create(context, pools):
+    """Create a pool from the values dictionary."""
+    session = get_session()
+    pool_refs = []
+    with session.begin():
+
+        for pool in pools:
+            LOG.debug('adding new pool for original_id {0}:'
+                      .format(pool.get('original_id')))
+            if not pool.get('id'):
+                pool['id'] = uuidutils.generate_uuid()
+
+            pool_ref = models.Pool()
+            pool_ref.update(pool)
+            pool_refs.append(pool_ref)
+
+        session.add_all(pool_refs)
+
+    return pool_refs
+
+
+def pools_delete(context, pools):
+    """Delete multiple pools with the pools dictionary."""
+    session = get_session()
+    pool_refs = []
+    with session.begin():
+        for pool in pools:
+            LOG.debug('deleting pool {0}:'.format(pool.get('id')))
+            query = _pool_get_query(context, session)
+            result = query.filter_by(id=pool.get('id')
+                                     ).delete()
+
+            if not result:
+                LOG.error(exception.PoolNotFound(id=pool.get(
+                    'id')))
+
+            pool_refs.append(pool)
+
+    return pool_refs
```

Comment:
Why is the return a list reference of pool

Reply:
It's not reference. Just another list of pools(which we got from storage).

Reply:
done